libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -1142,19 +1142,106 @@ genFromSpecT (simplifySpec -> spec) = case spec of
   where
     shr = shrinkWithTypeSpec @fn (emptySpec @fn @a)
 
+-- TODO: make sure we don't re-do unnecessary work for every shrink here?
+shrinkFromPreds :: HasSpec fn a => Pred fn -> Var a -> a -> [a]
+shrinkFromPreds p
+  | Result _ plan <- prepareLinearization p = \x a -> listFromGE $ do
+      -- NOTE: we do this to e.g. guard against bad construction functions in Exists
+      xaGood <- checkPred (singletonEnv x a) p
+      unless xaGood $
+        fatalError1 "Trying to shrink a bad value, don't do that!"
+      initialEnv <- envFromPred (singletonEnv x a) p
+      return
+        [ a'
+        | env' <- shrinkEnvFromPlan initialEnv plan
+        , Just a' <- [lookupEnv env' x]
+        , -- NOTE: this is necessary because it's possible that changing
+        -- a particular value in the env during shrinking might not result
+        -- in the value of `x` changing and there is no better way to know than
+        -- to do this.
+        a' /= a
+        ]
+  | otherwise = error "Bad pred"
+
+-- Start with a valid Env for the plan and try to shrink it
+shrinkEnvFromPlan :: Env -> SolverPlan fn -> [Env]
+shrinkEnvFromPlan initialEnv SolverPlan {..} = go mempty solverPlan
+  where
+    go _ [] = [] -- In this case we decided to keep every variable the same so nothing to return
+    go env ((substStage env -> SolverStage {..}) : plan) = do
+      Just a <- [lookupEnv initialEnv stageVar]
+      -- Two cases:
+      --  - either we shrink this value and try to fixup every value later on in the plan or
+      [ env' <> fixedEnv
+        | a' <- shrinkWithSpec stageSpec a
+        , let env' = extendEnv stageVar a' env
+        , Just fixedEnv <- [fixupPlan env' plan]
+        ]
+        --  - we keep this value the way it is and try to shrink some later value
+        ++ go (extendEnv stageVar a env) plan
+
+    -- Fix the rest of the plan given an environment `env` for the plan so far
+    fixupPlan env [] = pure env
+    fixupPlan env ((substStage env -> SolverStage {..}) : plan) =
+      case lookupEnv initialEnv stageVar >>= fixupWithSpec stageSpec of
+        Nothing -> Nothing
+        Just a -> fixupPlan (extendEnv stageVar a env) plan
+
+-- Try to fix a value w.r.t a specification
+fixupWithSpec :: forall fn a. HasSpec fn a => Specification fn a -> a -> Maybe a
+fixupWithSpec spec a
+  | a `conformsToSpec` spec = Just a
+  | otherwise = case spec of
+      MemberSpec (a :| _) -> Just a
+      _ -> listToMaybe $ filter (`conformsToSpec` spec) (shrinkWithSpec @fn TrueSpec a)
+
+-- Construct an environment for all variables that show up on the top level from an
+-- environment for all the free variables in the pred. The environment you get out of
+-- this function is _bigger_ than the environment you put in. From
+-- ```
+-- let y = x + 1 in let z = y + 1 in foo x y z
+-- ```
+-- and an environment with `{x -> 1}` you would get `{x -> 1, y -> 2, z -> 3}` out.
+envFromPred :: Env -> Pred fn -> GE Env
+envFromPred env p = case p of
```

Comment:
It looks like, only Let and Exists are the only Pred's that add anything, so it appears that these are the places free variables are bound, Might put a bit of explanation about this in the comment?